wordpress-mobile/gutenberg-mobile
# Issue/715 request cancel image upload

Fixes #715 

Gutenberg PR https://github.com/WordPress/gutenberg/pull/14391
WPAndroid PR https://github.com/wordpress-mobile/WordPress-Android/pull/9402

Until now, when the user started uploading an image in an image block, the image will continue to upload until finished doing so regardless of the block containing such image being deleted.

This PR adds a new method to the bridge interface to request the host app to handle an in-progress upload cancellation.

The bridge methods added are self-explained as they follow the same structure of what we already have in place, but the way we detect a block being deleted might use some further explanation:
- when the user taps on the trash icon, we detect that event without ambiguity [in the Toolbar handler](https://github.com/wordpress-mobile/gutenberg-mobile/compare/develop...issue/715-request-cancel-image-upload?expand=1#diff-2004b9687d71de0f0a9816269bb4a6a2R109). It is at this point that we ~[add a filter](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-hooks/)~ add an action that will be called later. The ~filter~ action is added here to handle the situation, but is needed to be invoked from the component that has knowledge of the `mediaId` corresponding to the in-progress upload.
- when the Image component that is part of this `ImageBlock` is [about to be unmounted](https://github.com/WordPress/gutenberg/pull/14391/files#diff-a61b3f346c11f42236eb2b1a783bf9bfR91), we make use of the action. _If_ the ~filter~ action exists, it will be executed and is passed the needed information (media `id` and whether it's an in progress upload or not).
- the ~filter~ action then [checks that the mediaId exists and is an in-progress upload](https://github.com/wordpress-mobile/gutenberg-mobile/compare/develop...issue/715-request-cancel-image-upload?expand=1#diff-2004b9687d71de0f0a9816269bb4a6a2R85) and signals the host app to cancel this upload by means of the bridge.

- Finally, the ~filter~ action is removed as it no longer serves a purpose. You can see this ~filter~ action is added and removed in a brief amount of time.

I'm not 100% sold on this approach so open to suggestions, but I couldn't find a better way to:
- a) make sure the intent from the user is detected unequivocally (that is, they really did tap on the trash icon, intending to eliminate this block entirely). This is where the filter is installed, so the Image component (who knows about mediaID) knows it's being unmounted because of a removal.
- b) get the media ID information from the underlying Image component (as this information lives there at that level, and not on the Block or Block Holder levels).
- c) be able to call the bridge from where it makes sense. Note that we could also do this at the Image component level (as we are doing with other Media elements), but it still made sense to make the filter perform the bridge call and remove itself afterwards, giving the loop a closure.

*Note:* The iOS side of the bridge still needs to be implemented

To test:
1. start a draft
2. insert an image block
3. choose an image from the device so it starts uploading
4. once it starts uploading, tap on the block's trash icon
5. observe the foreground upload notification is also dismissed (meaning, the upload has been cancelled).
